torchao/quantization/quant_primitives.py

```diff
@@ -323,21 +323,23 @@ def dynamically_quantize_per_tensor(
 
 
 def quantize_activation_per_token_absmax(t):
-    n_bits = 8
     # if the shape of t is [B, N, K], the shape of scales will be [B, N, 1]
-
-    scales = t.abs().amax(dim=-1, keepdim=True)
-    if scales.dtype == torch.float16:
-        scales = (
-            scales.float()
-        )  # want float scales to avoid overflows for fp16, (bf16 has wide enough range)
-    q_max = 2 ** (n_bits - 1) - 1
-    scales = scales.clamp(min=1e-5).div(q_max)
```

Comment:
@vkuzo @HDCharles do you know if this clampping before the scale is calculated is required for `quantize_activation_per_token_absmax`?

Reply:
we need an epsilon to handle the case where max(abs(x)) is zero

Reply:
I see, so can we do this after we divide by q_max? that's where we are doing the clamp typically

Reply:
```
    if eps is not None:
        scale = torch.clamp(scale, min=eps)
```

you have this logic in `choose_qparams_affine`, I would imagine it's for the same purpose.  I would recommend:
1. always specify epsilon (not sure why this is optional)
2. ensure test cases include testing for `max(abs(x)) == 0`

Reply:
@vkuzo Is `eps` best chosen to be (a multiple of / the) machine epsilon of the input tensor's dtype? Or is it a parameter that needs to be searched depending on input data distribution?

Reply:
OK I'll add a test. will think about if we want to make it required, we didn't do clamping in some of the ops right now I think

Reply:
> will think about if we want to make it required

IMO the case of `max(abs(x)) == 0` should be handled for every op

> Is eps best chosen to be (a multiple of / the) machine epsilon of the input tensor's dtype?

I think that choosing eps based on the dtype makes sense, it should not be data dependent.  My educated guess of how to choose this number would be "smallest number so that the resulting scale is finite", I haven't looked into whether expressing that in terms of machine epsilon would make sense.

Reply:
after a second thought I guess it makes sense to have an eps, otherwise we may have divide by zero during quantization, I'll change it in a separate PR